### PR TITLE
Expand test coverage for undertested core components

### PR DESCRIPTION
**Pull request by DominicBurkart**
## Summary

Adds `tests/core_coverage.rs` with targeted tests for gaps identified through coverage analysis of the core library. The existing tests are strong on the happy paths and main collision scenarios; this PR fills in edge cases, invariant checks, and under-exercised code paths.

### Channel edge cases (11 tests, 2 proptests)
- Empty channel operations: `resolve_at`, `drain_completed`, `deliver_to` on a fresh channel
- Partial resolution: only transmissions finished by the given time are resolved
- Direct `compute_rssi` / `compute_snr` validation (including negative TX power)
- `deliver_to` time filtering: completed transmissions beyond the query time are excluded
- Late-arriving strong signal capturing an earlier weak one (order reversal)
- Event variant assertions: `begin_transmission` returns `TransmissionStarted`, `resolve_at` returns `TransmissionCompleted` with correct fields
- Multiple resolve/drain cycles
- **Proptest**: capture asymmetry holds for arbitrary power deltas >= threshold
- **Proptest**: `resolve_at` before a transmission ends always returns empty

### Scheduler invariants (9 tests, 1 proptest)
- `current_time` progresses through scheduled wakes
- Events beyond `end_time` are correctly skipped
- Determinism: identical setup produces identical metrics and final time
- Multi-hop reply chain: A -> B -> A round-trip delivery
- Two simultaneous interferers with independent airtime accounting
- Interferer on different SF causes no collision with node TX
- Empty scheduler (no nodes, no events) finishes at time 0
- Per-node TX/RX count isolation through the scheduler
- **Proptest**: total airtime equals sum of individual TX durations (no-collision case)

### Metrics isolation (2 tests)
- Per-node TX/RX counters are independent across nodes
- `Default` and `new()` produce equivalent zero-state collectors

### Time conversion edge cases (2 tests, 1 proptest)
- `sim_time_to_ms` truncation for sub-millisecond values (999us -> 0ms, 1500us -> 1ms)
- `ms_to_sim_time` at `u32::MAX` boundary
- **Proptest**: `sim_time_to_ms` output * 1000 never exceeds input (truncation invariant)

## Test plan

- [ ] `cargo test --all-features` passes (all new and existing tests)
- [ ] `cargo clippy --all-features --all-targets` produces no warnings
- [ ] Coverage delta is positive on the core modules (channel, scheduler, metrics, time)
">

### CODE REVIEW

File: tests/core_coverage.rs
Comment (DominicBurkart) [inline tests/core_coverage.rs:1]: **Cross-PR conflict: #14 creates the same file.** PR #14 also adds `tests/core_coverage.rs` with ~607 lines of overlapping test coverage. These two PRs will conflict on merge. Since #17 appears to be a refined/expanded successor to #14 (adds time conversion tests, removes the `EventKind` equality tests that the #14 review flagged, drops `deliver_to` usage in favor of `drain_completed` in some places), the recommended path is:
1. Close #14 in favor of #17, or
2. Rebase #17 on top of #14 and make it purely additive.

Do not merge both -- the file-level conflict will require a full manual resolution.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+
+#[test]
+fn channel_empty_resolve_returns_nothing() {
+    let mut ch = Channel::new();
+    let events = ch.resolve_at(1_000_000);
+    assert!(events.is_empty());
+}
+
+#[test]
+fn channel_empty_drain_returns_nothing() {
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:174]: **Cross-PR conflict: #16 removes `deliver_to`.** This test calls `ch.deliver_to()` which PR #16 deletes. If #16 merges first, this test will fail to compile. Either:
- Coordinate merge order (#17 before #16), or
- Remove the `deliver_to` calls from this PR now, since #16 makes a good case that `deliver_to` is dead production code with subtly wrong semantics (non-draining reads).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 2]
```diff
@@ -0,0 +1,547 @@
...
+
+#[test]
+fn channel_empty_deliver_returns_nothing() {
+    let ch = Channel::new();
+    assert!(ch.deliver_to(1_000_000).is_empty());
+}
+
+#[test]
+fn channel_resolve_partial_only_finished() {
+    let mut ch = Channel::new();
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:181]: This test couples to `Channel::new()` default constants (`path_loss_db = 100.0`, `noise_floor_dbm = -117.0`). If those defaults change, this test breaks silently with a confusing assertion. Consider naming the test to make the coupling explicit (e.g., `channel_default_rssi_and_snr_with_known_constants`) or extracting the expected values from the channel's public API if available.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 3]
```diff
@@ -0,0 +1,547 @@
...
+    sched.add_node(Box::new(node), Some(0));
+    sched.run();
+    assert!(sched.current_time() <= 100_000);
+}
+
+#[test]
+fn scheduler_determinism() {
+    fn run_scenario() -> (u64, u64, u64, u64) {
+        let mut sched = Scheduler::new(500_000);
+        let mut s1 = TestNode::new(1);
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:375]: Good test. The `EchoNode` helper is well-designed for validating multi-hop delivery through the scheduler. This is one of the highest-value additions in the PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 4]
```diff
@@ -0,0 +1,547 @@
...
+}
+
+#[test]
+fn scheduler_two_interferers_simultaneously() {
+    let mut sched = Scheduler::new(300_000);
+    sched.add_node(Box::new(TestNode::new(1)), None);
+
+    let i1 = CountingInterferer {
+        tx: tx(7, 868_100_000, 30_000, 14),
+        remaining: 2,
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:414]: Nit: The comment says "node TX + interferer TX delivered on different SFs" but the assertion value of 3 deserves a breakdown in the comment. Something like: "Node1 SF7 -> Node2 (1 RX) + Interferer SF12 -> Node1 and Node2 (2 RX) = 3 total".

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by DominicBurkart — COMMENTED** [review]
Solid test suite expansion. The scheduler invariant tests (determinism, multi-hop echo, interferer isolation) and time conversion edge cases are the highest-value additions. The channel edge-case tests are thorough.

**Blocking cross-PR issues:**

1. **Conflicts with PR #14** -- both create `tests/core_coverage.rs`. This PR appears to be a refined successor (#17 drops the `EventKind` tests flagged in the #14 review, adds time conversion and interferer tests). Recommend closing #14 in favor of this PR.

2. **Conflicts with PR #16** -- this PR calls `ch.deliver_to()` in 3 tests, but #16 removes that method entirely. Decide on merge order or drop `deliver_to` usage here preemptively.

**Non-blocking:**
- RSSI/SNR exact-value assertions couple to `Channel::new()` default constants. Name the tests to make this explicit.
- The `assert_eq!(sched.metrics.total_rx, 3, ...)` comment could be more precise about the RX breakdown.

Overall this is a clear improvement to test coverage. The cross-PR coordination is the main blocker.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+    let ch = Channel::new();
+    assert!(ch.deliver_to(1_000_000).is_empty());
+}
+
+#[test]
+fn channel_resolve_partial_only_finished() {
+    let mut ch = Channel::new();
+    // TX1 ends at 50_000, TX2 ends at 150_000
+    ch.begin_transmission(NodeId(1), &tx(7, 868_100_000, 50_000, 14), 0);
+    ch.begin_transmission(NodeId(2), &tx(8, 868_100_000, 150_000, 14), 0);
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:184]: **blocking: conflict with PR #16**

`deliver_to` is called in this test, `channel_deliver_to_filters_by_time`, and `capture_is_asymmetric`. PR #16 removes `Channel::deliver_to()` entirely. If #16 merges first, these three tests will fail to compile.

Recommended fix: replace all `deliver_to` usage with `drain_completed`, which is the production-path API the scheduler actually uses. This also eliminates testing dead code.

If this PR is intended to supersede #16 and keep `deliver_to`, document that decision explicitly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+            None
+        }
+    }
+}
+
+// ===========================================================================
+// Channel edge cases
+// ===========================================================================
+
+#[test]
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:166]: **nit: assertion couples to default `Channel::new()` constants**

`-86.0` and `31.0` are derived from `path_loss_db=100.0` and `noise_floor_dbm=-117.0`, which are private defaults inside `Channel::new()`. If those defaults change, this test silently tests the wrong thing rather than failing obviously.

Consider either:
1. Renaming the test to `channel_compute_rssi_with_default_path_loss` to make the coupling explicit, or
2. Computing expected values from named constants: `let expected_rssi = 14.0 - 100.0;` with a comment stating the source.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+    ch.resolve_at(200_000);
+
+    // deliver_to at 60_000 should only include TX1
+    let delivered = ch.deliver_to(60_000);
+    assert_eq!(delivered.len(), 1);
+    assert_eq!(delivered[0].sf, 7);
+
+    // deliver_to at 200_000 should include both
+    let delivered_all = ch.deliver_to(200_000);
+    assert_eq!(delivered_all.len(), 2);
```
Comment (DominicBurkart) [inline tests/core_coverage.rs]: **nit: proptest `capture_is_asymmetric` tests `deliver_to` count but not *which* signal survived**

The assertion `delivered.len() == 1` confirms capture occurred, but does not verify the *strong* signal is the survivor. A captured weak signal that somehow passed through would also satisfy `len() == 1`. Consider adding:
```rust
prop_assert_eq!(delivered[0].rssi, ch.compute_rssi(strong_power));
```
(This also applies to the non-proptest `channel_late_strong_captures_earlier_weak` test at line 195, which correctly checks via `drain_completed` tuple fields -- good.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+
+impl EchoNode {
+    fn new(id: u32, reply: Transmission) -> Self {
+        Self {
+            id: NodeId(id),
+            reply: Some(reply),
+            received: Vec::new(),
+        }
+    }
+}
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:114]: **suggestion: `EchoNode::poll_transmit` has a subtle one-shot limitation worth documenting**

`poll_transmit` returns `self.reply.take()` only when `self.received` is non-empty, but after the first `take()` it will always return `None` regardless of further receives. This is correct for the single-echo test but could confuse future test authors who reuse `EchoNode` expecting multi-reply behavior.

Add a doc comment like `/// Replies exactly once, on the first poll after the first receive.`

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+#[test]
+fn scheduler_determinism() {
+    fn run_scenario() -> (u64, u64, u64, u64) {
+        let mut sched = Scheduler::new(500_000);
+        let mut s1 = TestNode::new(1);
+        s1.pending_tx = Some(tx(7, 868_100_000, 50_000, 14));
+        let mut s2 = TestNode::new(2);
+        s2.pending_tx = Some(tx(7, 868_100_000, 50_000, 14));
+        sched.add_node(Box::new(s1), Some(0));
+        sched.add_node(Box::new(s2), Some(10_000));
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:380]: **nit: assertion comment could be more precise**

The comment `"node TX + interferer TX delivered on different SFs"` is correct but the RX breakdown is non-obvious. The previous review flagged this too. Consider:
```rust
// Node 1 TX (SF7) -> delivered to Node 2 (1 RX)
// Interferer TX (SF12) -> delivered to Node 1 and Node 2 (2 RX)
// Total: 3 RX
assert_eq!(sched.metrics.total_rx, 3);
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+        interval: 80_000,
+    };
+    sched.add_interferer(Box::new(i1), 0);
+    sched.add_interferer(Box::new(i2), 0);
+    sched.run();
+
+    // Total airtime = 2*30_000 + 3*20_000 = 120_000
+    assert_eq!(sched.metrics.total_airtime_us, 120_000);
+}
+
```
Comment (DominicBurkart) [inline tests/core_coverage.rs:429]: **suggestion: proptest `scheduler_airtime_equals_sum_of_durations` could be strengthened**

The two TXs use different SFs (7 and 8), which avoids collision, but they both start at `Some(0)` -- meaning they overlap in time. The test relies on different-SF isolation. This is fine, but would be more robust if the test name or a comment made this explicit: "non-colliding because different SFs, not because non-overlapping."

Also consider adding a case where the TXs are on the *same* SF but non-overlapping, to verify airtime accounting is correct independent of collision status.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by DominicBurkart — COMMENTED** [review]
## Automated Review — PR #17: Expand test coverage for undertested core components

**Verdict: CHANGES REQUESTED** (1 blocking issue, 5 non-blocking suggestions)

### Overall assessment

This is a well-structured test expansion. The 25 tests + 4 proptests cover meaningful gaps: empty-channel edge cases, partial resolution, scheduler determinism, multi-hop echo, interferer isolation, metrics independence, and time conversion boundaries. Test naming is clear, helper structs are well-designed, and proptest strategies are appropriately constrained. No secrets, PII, or unnecessary bloat detected.

### Blocking

1. **`deliver_to` usage conflicts with PR #16.** Three tests (`channel_empty_deliver_returns_nothing`, `channel_deliver_to_filters_by_time`, `capture_is_asymmetric` proptest) call `Channel::deliver_to()`, which PR #16 removes. If #16 merges first, this PR will not compile. Either migrate these to `drain_completed` or explicitly coordinate merge order with #16. PR #14 (the predecessor to this PR) is already closed, so that conflict is resolved.

### Non-blocking

2. RSSI/SNR exact-value assertions are coupled to private `Channel::new()` defaults -- make this coupling explicit via naming or computed expected values.
3. `capture_is_asymmetric` proptest asserts count but not identity of the surviving signal.
4. `EchoNode` one-shot reply semantics should be documented to prevent misuse in future tests.
5. `total_rx == 3` assertion in `scheduler_interferer_and_node_different_sf_no_collision` deserves a per-node RX breakdown comment.
6. `scheduler_airtime_equals_sum_of_durations` relies on different-SF isolation without stating so; consider adding a same-SF non-overlapping variant.

### Gaps not covered by this PR

- No tests for `Channel::with_co_channel_rejection()` with the new test helpers (existing unit tests cover this, but the integration-level helpers could exercise it too).
- No negative test for `ms_to_sim_time(0)` (trivial but would round out the edge cases).

### CI Status

All 6 check runs passed on commit `a670218`: Check, Test, Code Coverage, Cargo Deny, Autoformat, Security Audit.

### Recommendation

Not ready for human merge review yet. Resolve the `deliver_to` / PR #16 conflict (the single blocking item), then this is good to go. The non-blocking items can be addressed in this PR or deferred.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by DominicBurkart — COMMENTED** [review]
## Code Review — PR #17: Expand test coverage for undertested core components

**Verdict: COMMENT** — No new blockers beyond what the prior review identified. Several test-correctness issues worth addressing before merge: one proptest that cannot catch a meaningful failure class, one proptest reliance on an undeclared assumption, and one fragile assertion pattern. Details below.

---

### Test correctness

**`capture_is_asymmetric` proptest asserts count but not identity**

The proptest generates a strong signal on NodeId(1) and a weak signal on NodeId(2) and then asserts `delivered.len() == 1`. This passes even if the *wrong* signal survives — e.g. a bug that made the weak signal the survivor would not be caught because the count is still 1. A stronger assertion would verify the surviving frame carries the strong signal's SF/power characteristics or check `drain_completed` to confirm NodeId(1) is not collided:

```rust
let completed = ch.drain_completed();
let strong_entry = completed.iter().find(|(id,_,_,_)| *id == NodeId(1)).unwrap();
prop_assert!(!strong_entry.1, "strong signal must not be collided");
prop_assert!(strong_entry.2, "strong signal must be marked captured");
```

---

**`scheduler_airtime_equals_sum_of_durations` does not verify no collision occurred**

The two nodes use SF7 and SF8 respectively, which prevents channel-level collision, but the test never asserts `total_collisions == 0`. If a future scheduler change mistakenly collided different-SF transmissions, the airtime accumulation would still pass (`record_airtime` fires unconditionally in `handle_poll_transmit` regardless of collision outcome), making this a vacuous invariant check for the collision-freedom property it implies. Adding `prop_assert_eq!(sched.metrics.total_collisions, 0)` would close this.

---

**`channel_compute_rssi_and_snr_directly` — opaque magic numbers**

The test asserts `rssi == -86.0` and `snr == 31.0` derived silently from `Channel::new()` defaults (`path_loss_db=100.0`, `noise_floor_dbm=-117.0`). If those defaults ever change, the failure message gives no clue why. Either name the expected values with a comment (`// 14 - 100 = -86.0 using Channel::new() defaults`) or compute them:

```rust
let expected_rssi = 14.0_f32 - 100.0; // tx_power - path_loss
let expected_snr = expected_rssi - (-117.0); // rssi - noise_floor
```

This matches the pattern already used in `rssi_derived_from_tx_power` in the unit tests.

---

### Proptest quality

**`resolve_at_before_end_returns_nothing` — `dur/2` truncation can produce `midpoint == end`**

When `dur == 1`, `dur / 2 == 0`, so `midpoint = start + 0 = start`. The transmission ends at `start + 1`. `resolve_at(start)` checks `tx.end <= time` → `start+1 <= start` = false, so the proptest still passes. However the comment says "resolve before transmission ends" and the test name says "before end" — the tested time is actually the *start* time, not a midpoint. This is fine functionally but the intent is misleading. Using `start` directly as the resolve time would be cleaner and the same result:

```rust
let events = ch.resolve_at(start); // before tx ends at start+dur
```

---

### Coverage gaps not addressed by this PR

1. `Channel::with_co_channel_rejection(db)` — covered in unit tests but absent from the integration test helpers. The `capture_is_asymmetric` proptest implicitly relies on the default 6.0 dB threshold; a companion test using `with_co_channel_rejection` would confirm the configurable path is exercised at the integration level.

2. `EchoNode` is inherently one-shot (the `reply` field is consumed on first `poll_transmit`). The multi-hop test covers A→B→A but not A→B→C or longer chains. This is fine for the stated scope but worth noting if the PR description's "multi-hop reply chain" is intended to cover the general case.

3. `scheduler_airtime_equals_sum_of_durations` tests the no-collision path but has no companion test for the collision path — i.e., that airtime *still* accumulates correctly when a TX is collided (since `record_airtime` fires before collision resolution). The existing unit test `active_interferer_injects_and_records_airtime` in `scheduler.rs` covers part of this, but an explicit integration-level case would be clearer.

---

### Non-issues confirmed

- `channel_late_strong_captures_earlier_weak`: capture semantics are correct against the source — strong arrives at t=10_000, overlap is detected against the weak TX [0,80_000), delta=12≥6, weak gets `collided=true`, strong gets `captured=true`. Assertions match.
- `scheduler_interferer_and_node_different_sf_no_collision` `total_rx==3` assertion: node 1's SF7 TX delivers to node 2 (1 RX), then interferer's SF12 TX delivers to both node 1 and node 2 (2 RX) = 3 total. Arithmetic is correct.
- `scheduler_two_interferers_simultaneously` airtime=120_000: i1 fires 2×30_000=60_000, i2 fires 3×20_000=60_000, all within end_time=300_000. Correct.
- `metrics_default_is_same_as_new`: `MetricsCollector::new()` delegates to `Self::default()`, so the equality holds by construction.
- `ms_to_sim_time_max_u32`: `u32::MAX as u64 * 1_000` fits in u64 (≈4.29×10¹²). No overflow.

---

### Summary

The previous review's blocking item (`deliver_to` vs PR #16) remains the only thing preventing compile-clean merge. Among test-correctness issues: the `capture_is_asymmetric` proptest is the most actionable — it tests a meaningful property but the assertion form lets a whole failure class (wrong survivor) slip through silently. The others are low-severity polish items.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by DominicBurkart — COMMENTED** [review]
## Re-check on head b856cce — COMMENT (no new issues)

`tests/core_coverage.rs` content is byte-for-byte identical to what the Apr 5 review analyzed (`3f1f0cd`). The three new commits appear to be rebase/sync work; no test logic changed. All 6 CI checks pass on the new head.

**Status of every prior finding:**

| Finding | Disposition |
|---|---|
| `deliver_to` removed by PR #16 — compile blocker | **Still open.** The three `deliver_to` call sites in this file remain. Verify whether #16 has merged to main; if so, this PR will not compile without migrating those calls to `drain_completed`. |
| `capture_is_asymmetric` asserts count but not identity of survivor | Still open — the `rssi` field assertion added in the current file does check identity indirectly (RSSI must match strong sender's power), so this is **partially addressed**. A full closure would also assert `survivors[0].0 == NodeId(1)` (the strong sender's NodeId). |
| `resolve_at_before_end_returns_nothing` — `dur/2` truncation means midpoint can equal start when `dur==1` | Still open — functionally harmless but intent is misleading. |
| `scheduler_airtime_equals_sum_of_durations` — no `total_collisions == 0` assertion | Still open. |
| `channel_compute_rssi_negative_power` — expected value `-110.0` is an opaque magic number (compare: the RSSI/SNR test now computes expected values explicitly) | Inconsistency remains; minor. |
| `EchoNode` one-shot semantics undocumented | Addressed — doc comment is present in the file. |
| `total_rx == 3` comment in `scheduler_interferer_and_node_different_sf_no_collision` | Addressed — inline comment breaks down the RX count. |

**One new observation on the current file:**

`channel_late_strong_captures_earlier_weak` resolves at `t=80_000`, which is exactly when the weak TX ends (`start=0, dur=80_000`). Whether `resolve_at` uses `end <= time` or `end < time` determines whether the weak TX is included in this call. The test relies on the weak TX being included (it asserts `weak_entry.1 == true`). This is correct under `end <= time` semantics (which the prior review confirmed), but the test name and setup silently depend on that boundary condition. Renaming to `channel_late_strong_captures_earlier_weak_at_boundary` or adding a comment `// resolve_at uses end <= time, so both TXs resolve here` would make the boundary dependency explicit rather than accidental.

**Recommended actions before removing DRAFT status:**

1. Confirm PR #16 merge status and resolve `deliver_to` vs `drain_completed` accordingly (blocking for compile).
2. Add `prop_assert_eq!(survivors[0].0, NodeId(1))` to `capture_is_asymmetric` to close the survivor-identity gap.
3. Add `prop_assert_eq!(sched.metrics.total_collisions, 0)` to `scheduler_airtime_equals_sum_of_durations`.

Items 2 and 3 are each a one-line addition; item 1 is the only blocker.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by DominicBurkart — COMMENTED** [review]
[Automated review by Claude] — re-review on head `b856cce`

## Verdict: COMMENT — no blockers; three small polish items remain

The PR is a clean, high-value test expansion. 100% coverage now reported by Codecov, all 6 CI checks green, and the previously-flagged blocking issue (`deliver_to` removal in PR #16) is fully resolved — every test in `tests/core_coverage.rs` uses `resolve_at`/`drain_completed` exclusively. Several prior-review items have also been addressed in-file: RSSI/SNR expected values are now computed from the documented coupling, the `total_rx == 3` interferer test has an explicit per-source RX breakdown, and `EchoNode::poll_transmit` has a one-shot semantics docstring.

### Test quality assessment

These are real coverage gains, not tests-of-tests:

- **Channel edge cases** (empty/partial/multi-cycle resolve, captured/collided ordering, event-variant assertions) target genuinely under-exercised paths in `Channel`. Assertions verify behavior, not just non-panic.
- **Scheduler invariants** — determinism (`run_scenario() == run_scenario()`), multi-hop echo, two-interferer airtime accounting, different-SF non-collision, per-node TX/RX isolation, and end-time skip — are precisely the kinds of structural properties unit tests in `scheduler.rs` cannot easily express. The `scheduler_no_nodes_no_events_finishes_immediately` case is trivial but cheap and worth keeping as a regression guard.
- **Time conversion proptest** (`sim_time_to_ms_never_exceeds_input`) is a meaningful truncation invariant, not a tautology.
- **Metrics isolation** tests are minimal but appropriate; `metrics_default_is_same_as_new` is essentially a constructor equivalence check (low signal — `new()` literally calls `default()` per the source) but adds negligible bloat.

No flakiness risk: no clock/randomness/threading; proptest seeds are deterministic; ranges are bounded.

### Remaining non-blocking polish (carried from prior review)

1. **`capture_is_asymmetric` proptest** — now asserts the survivor's RSSI matches the strong sender's expected RSSI, which closes most of the survivor-identity gap. Adding `prop_assert_eq!(survivors[0].0, NodeId(1))` would make the identity check explicit and resilient to any future change that decouples RSSI from sender identity.
2. **`scheduler_airtime_equals_sum_of_durations`** — the comment now correctly notes different-SF orthogonality, but adding `prop_assert_eq!(sched.metrics.total_collisions, 0)` would make the no-collision precondition a checked invariant rather than an assumption (`record_airtime` fires unconditionally, so the airtime equality alone does not detect a future regression that incorrectly collides different-SF TXs).
3. **`resolve_at_before_end_returns_nothing`** — `dur/2` truncation means `midpoint == start` when `dur == 1`. Functionally correct but mildly inconsistent with the test name ("before end"). Either bump the lower bound to `dur in 2u64..` or use `start` directly with a clarifying comment.

### Boundary observation

`channel_late_strong_captures_earlier_weak` resolves at `t=80_000`, which equals the weak TX's end time. The assertion that `weak_entry.1 == true` depends on `resolve_at` using `end <= time` (inclusive). This is the source semantics, but the dependency is implicit — a one-line comment (`// resolve_at uses end <= time, so the weak TX (ends at 80_000) is included`) would prevent confusion if the boundary semantics ever change.

### Coverage gaps not addressed (appropriate to defer)

- `Channel::with_co_channel_rejection()` is unit-tested but not exercised at integration level via the new helpers.
- `EchoNode` only models A→B→A; A→B→C requires a different helper (one-shot reply consumed on first poll).
- No test for the collision-path airtime accumulation (i.e., that `record_airtime` correctly fires before collision resolution); `active_interferer_injects_and_records_airtime` partially covers this.

### Recommendation

Ready for human merge review once items 1–3 above are addressed (each is a one-line change). The PR is otherwise mergeable as-is — none of the remaining items rise to "blocking" given the 100% coverage and clean CI.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 1 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:174]: Addressed in b856cce: all `deliver_to` calls have been replaced with `drain_completed`, eliminating the conflict with PR #16 and avoiding tests against dead code.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 2 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:181]: Addressed in b856cce: test renamed to `channel_compute_rssi_and_snr_with_default_path_loss` and a comment was added explaining the coupling to `Channel::new()` defaults (`path_loss_db=100.0`, `noise_floor_dbm=-117.0`) and how the expected values are derived.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 4 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:414]: Addressed in b856cce: an inline RX breakdown was added above the assertion (`Node 1 TX (SF7) -> Node 2 = 1 RX; Interferer (SF12) -> Node 1 + Node 2 = 2 RX; Total = 3`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 5 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:184]: Addressed in b856cce: all three `deliver_to` call sites were rewritten to use `drain_completed`. Compile-conflict with PR #16 is resolved.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 6 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:166]: Addressed in b856cce: test renamed to `channel_compute_rssi_and_snr_with_default_path_loss`; expected values are now computed from named literals (`14.0 - 100.0`, `rssi - (-117.0)`) with a comment citing the source defaults.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 7 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs]: Addressed in b856cce: `capture_is_asymmetric` proptest now switches to `drain_completed` and asserts `survivors[0].3.rssi == ch.compute_rssi(strong_power)` to verify the strong signal is the actual survivor.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 8 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:114]: Addressed in b856cce: added doc comment `/// Replies exactly once, on the first poll after the first receive.` on `EchoNode::poll_transmit`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 9 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:380]: Addressed in b856cce: the breakdown comment now explicitly enumerates `Node 1 TX (SF7) -> Node 2 (1 RX)` and `Interferer (SF12) -> Node 1 and Node 2 (2 RX)` totaling 3.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 10 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:429]: Partially addressed in b856cce: a comment was added clarifying that non-collision is due to different-SF orthogonality, not non-overlapping windows. The additional same-SF non-overlapping case is left for a follow-up to keep this PR focused.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs
Comment (DominicBurkart) [reply tests/core_coverage.rs:1]: Cross-PR coordination note acknowledged. This PR is the refined successor to #14 and the deliver_to conflict with #16 has now been resolved (see b856cce, which replaces all `deliver_to` calls with `drain_completed`). Recommend closing #14 in favor of this PR; merge order is no longer constrained relative to #16.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 3 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:375]: Thanks - noted, no changes needed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by DominicBurkart** [comment]
## Address remaining review polish — commit 1d101a8

Three items flagged as still-open in the Apr 14 and Apr 16 reviews are now addressed in commit `1d101a8`:

**1. `capture_is_asymmetric` — explicit survivor NodeId assertion (was: count + RSSI only)**
Added `prop_assert_eq!(survivors[0].0, NodeId(1))` directly before the RSSI check. The strong sender's NodeId is now verified explicitly, closing the gap where a bug that produced the right RSSI from the wrong node would have passed.

**2. `scheduler_airtime_equals_sum_of_durations` — checked `total_collisions == 0`**
Added `prop_assert_eq!(sched.metrics.total_collisions, 0)` before the airtime equality assertion. Non-collision is now an enforced precondition rather than an implicit assumption, so a future regression that incorrectly collides different-SF TXs will be caught here.

**3. `resolve_at_before_end_returns_nothing` — proptest range tightened from `1u64..` to `2u64..`**
When `dur == 1`, `dur / 2 == 0` and `midpoint == start`, which tests resolution at the start time rather than a genuine midpoint. Tightening to `dur in 2u64..1_000_000` guarantees `dur / 2 >= 1` and the midpoint is always strictly between start and end. A clarifying comment was added to make this invariant explicit.

**Boundary note also addressed:**
`channel_late_strong_captures_earlier_weak` now has a comment documenting that `resolve_at` uses `end <= time` (inclusive) semantics, explaining why both TXs resolve at `t=80_000` and making the boundary dependency explicit.

No other logic was changed. All previously-addressed items remain intact.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 7 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs]: Fully addressed in 1d101a8: `capture_is_asymmetric` now asserts `survivors[0].0 == NodeId(1)` (explicit strong-sender NodeId) in addition to the RSSI check already present. A bug that made the wrong node the survivor can no longer pass this test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/core_coverage.rs (hunk 10 above)
Comment (DominicBurkart) [reply tests/core_coverage.rs:429]: Addressed in 1d101a8: added `prop_assert_eq!(sched.metrics.total_collisions, 0)` before the airtime equality assertion. The no-collision assumption is now a checked invariant — a future regression that incorrectly collides different-SF transmissions will be caught here rather than passing silently through the airtime check.

---
_Generated by [Claude Code](https://claude.ai/code)_
